lib/src/wizard.dart

```diff
@@ -189,27 +190,46 @@ class Wizard extends StatefulWidget {
 }
 
 class _WizardState extends State<Wizard> {
-  List<WizardRouteSettings> _routes = [];
+  late WizardController _controller;
 
   @override
   void initState() {
     super.initState();
-    _ensureInitialRoute();
+    _controller = widget.controller ??
+        WizardController(
+            routes: widget.routes!, initialRoute: widget.initialRoute);
   }
 
   @override
-  void didUpdateWidget(Wizard oldWidget) {
-    super.didUpdateWidget(oldWidget);
-    _routes.removeWhere((r) => !widget.routes.containsKey(r.name));
-    _ensureInitialRoute();
+  void dispose() {
+    if (widget.controller == null) {
+      _controller.dispose();
+    }
+    super.dispose();
   }
 
-  void _ensureInitialRoute() {
-    if (_routes.isEmpty) {
-      _routes = <WizardRouteSettings>[
-        WizardRouteSettings(
-            name: widget.initialRoute ?? widget.routes.keys.first),
-      ];
+  @override
+  void didUpdateWidget(Wizard oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    if (widget.controller != oldWidget.controller ||
+        !listEquals(
+            widget.routes?.keys.toList(), oldWidget.routes?.keys.toList()) ||
+        widget.initialRoute != oldWidget.initialRoute) {
```

Comment:
let's move this cheap initialRoute check before the more expensive list comparison. i think there's also a cheaper iterable equality check somewhere in dart:collection or package:collection to avoid having to convert the lazy key iterables to lists first